internal/pkg/source/source.go

```diff
@@ -75,104 +75,185 @@ func New(in ssa.Node, config classifier) *Source {
+		// (If the call is an operand, then it is being used as a value, so it does
+		// not matter when the call occurred.)
+		if _, ok := instr.(*ssa.Call); ok && instrIndex < maxInstrReached[instr.Block()] && isReferrer {
+			return true
+		}
 	}
+
+	return false
 }
 
-// referrersToVisit produces a filtered list of Referrers for an ssa.Node.
-// Specifically, we want to avoid referrers that shouldn't be visited, e.g.
-// because they would not be reachable in an actual execution of the program.
-func (s *Source) referrersToVisit(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int) (referrers []ssa.Instruction) {
-	if n.Referrers() == nil {
+func (s *Source) visit(n ssa.Node, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	if s.node == n {
+		s.visitReferrers(n, maxInstrReached, lastBlockVisited)
 		return
 	}
-	for _, r := range *n.Referrers() {
-		if c, ok := r.(*ssa.Call); ok {
-			// This is to avoid attaching calls where the source is the receiver, ex:
-			// core.Sinkf("Source id: %v", wrapper.Source.GetID())
-			if recv := c.Call.Signature().Recv(); recv != nil && s.config.IsSourceType(utils.DecomposeType(utils.Dereference(recv.Type()))) {
-				continue
-			}
 
-			// If this call's index is lower than the highest in its block,
-			// then this call is "in the past" and we should stop traversing.
-			i, ok := indexInBlock(r)
-			if !ok {
-				continue
-			}
-			if i < maxInstrReached[r.Block()] {
-				continue
-			}
+	switch t := n.(type) {
+	case *ssa.Alloc:
+		// An Alloc represents the allocation of space for a variable. If a Node is an Alloc,
+		// and the thing being allocated is not an array, then either:
+		// a) it is a Source value, in which case it will get its own traversal when sourcesFromBlocks
+		//    finds this Alloc
+		// b) it is not a Source value, in which case we should not visit it.
+		// However, if the Alloc is an array, then that means the source that we are visiting from
+		// is being placed into an array, slice or varags, so we do need to keep visiting.
+		if _, isArray := utils.Dereference(t.Type()).(*types.Array); isArray {
+			s.visitReferrers(n, maxInstrReached, lastBlockVisited)
 		}
 
-		if fa, ok := r.(*ssa.FieldAddr); ok {
-			deref := utils.Dereference(fa.X.Type())
-			typPath, typName := utils.DecomposeType(deref)
-			fieldName := utils.FieldName(fa)
+	case *ssa.Call:
+		if callee := t.Call.StaticCallee(); callee != nil && s.config.IsSanitizer(utils.DecomposeFunction(callee)) {
+			s.sanitizers = append(s.sanitizers, &sanitizer.Sanitizer{Call: t})
+		}
 
-			if !s.config.IsSourceField(typPath, typName, fieldName) {
-				continue
-			}
+		// This is to avoid attaching calls where the source is the receiver, ex:
+		// core.Sinkf("Source id: %v", wrapper.Source.GetID())
+		if recv := t.Call.Signature().Recv(); recv != nil && s.config.IsSourceType(utils.DecomposeType(utils.Dereference(recv.Type()))) {
+			return
+		}
```

Comment:
Good question. This is actually one of the more significant architectural flaws in the codebase right now, IMO.

The reason is that `GetSecret` will be identified as a `fieldpropagator`, which will be found in `levee` when checking for `Calls`, which will start a new traversal with the result of `GetSecret` as a source. `fieldpropagators` should be used in the core traversal logic, not anywhere else. I had a PR for that a long time ago, #77, but it got lost in the weeds. I think I ended up closing it because I wanted the `fieldpropagators` to be consumed by `sourcetype` or something of the sort.

Currently the propagation logic in `fieldpropagator` is not nearly as developed as that in `source`, and in fact I believe that logic should probably be shared: #130.

Reply:
Gotcha.  And yeah, you know opinion on the pipeline of analyzers here.  If these were unified in `levee` instead of propagation happening all in `source`... but that's a conversation for a future PR.